minidump-processor/src/stackwalker/mips.rs

```diff
@@ -0,0 +1,396 @@
+    // for all frames except the topmost one.
+    // There is no way of knowing if topmost frame belongs to a leaf or
+    // a nonleaf function.
+    if callee.trust != FrameTrust::Context {
+        last_sp += MIN_ARGS * POINTER_WIDTH;
+        count -= MIN_ARGS;
+    }
+
+    for i in 0..count {
+        let address_of_pc = last_sp.checked_add(i * POINTER_WIDTH)?;
+        let caller_pc = stack_memory.get_memory_at_address_be(address_of_pc)?;
+        if instruction_seems_valid(caller_pc, modules, symbol_provider).await {
+            // pc is pushed by CALL, so sp is just address_of_pc + ptr
+            let caller_sp = address_of_pc.checked_add(POINTER_WIDTH)?;
+
+            // Don't do any more validation, and don't try to restore fp
+            // (that's what breakpad does!)
+
+            trace!(
+                "unwind: scan seems valid -- caller_pc: {caller_pc:#08x}, caller_sp: {caller_sp:#08x}"
+            );
+
+            let mut caller_ctx = MipsContext::default();
+            caller_ctx.set_register(PROGRAM_COUNTER, caller_pc - (2 * POINTER_WIDTH));
+            caller_ctx.set_register(STACK_POINTER, caller_sp);
+
+            let mut valid = HashSet::new();
+            valid.insert(PROGRAM_COUNTER);
+            valid.insert(STACK_POINTER);
+
+            let context = MinidumpContext {
+                raw: MinidumpRawContext::Mips(caller_ctx),
+                valid: MinidumpContextValidity::Some(valid),
+            };
+            return Some(StackFrame::from_context(context, FrameTrust::Scan));
+        }
+    }
+
+    None
+}
+
+async fn instruction_seems_valid<P>(
+    instruction: Pointer,
+    modules: &MinidumpModuleList,
+    symbol_provider: &P,
+) -> bool
+where
+    P: SymbolProvider + Sync,
+{
+    if instruction < 0x1000 {
+        return false;
+    }
+
+    super::instruction_seems_valid_by_symbols(instruction as u64, modules, symbol_provider).await
+}
+
+#[async_trait::async_trait]
+impl Unwind for MipsContext {
+    async fn get_caller_frame<P>(
+        &self,
+        callee: &StackFrame,
+        grand_callee: Option<&StackFrame>,
+        stack_memory: Option<&MinidumpMemory<'_>>,
+        modules: &MinidumpModuleList,
+        _system_info: &SystemInfo,
+        syms: &P,
+    ) -> Option<StackFrame>
+    where
+        P: SymbolProvider + Sync,
+    {
+        let ctx = Mips32Context::try_from(self.clone());
+        let stack = stack_memory.as_ref()?;
+
+        // .await doesn't like closures, so don't use Option chaining
+        let mut frame = None;
+        if frame.is_none() {
+            match &ctx {
+                Ok(mips32) => {
+                    frame =
+                        get_caller_by_cfi(mips32, callee, grand_callee, stack, modules, syms).await
+                }
+                Err(mips64) => {
+                    frame =
+                        get_caller_by_cfi(mips64, callee, grand_callee, stack, modules, syms).await
+                }
+            }
+        }
```

Comment:
Frame pointers are unreliable trash on mips?

Reply:
yes, I don’t think they are being used at all for that purpose.